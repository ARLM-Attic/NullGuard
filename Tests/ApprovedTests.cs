#if(DEBUG)

using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.Build.Utilities;
using NUnit.Framework;

[TestFixture]
[UseReporter(typeof(DiffReporter))]
public class ApprovedTests
{
    [Test]
    public void ClassWithBadAttributes()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "ClassWithBadAttributes"));
    }

    [Test]
    public void ClassWithPrivateMethod()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "ClassWithPrivateMethod"));
    }

    [Test]
    public void GenericClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "GenericClass"));
    }

    [Test]
    public void InterfaceBadAttributes()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "InterfaceBadAttributes"));
    }

    [Test]
    public void SimpleClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "SimpleClass"));
    }

    [Test]
    public void SpecialClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "SpecialClass"));
    }

    [Test]
    public void UnsafeClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "UnsafeClass"));
    }

    private static string Decompile(string assemblyPath, string identifier = "")
    {
        var exePath = GetPathToILDasm();

        if (!string.IsNullOrEmpty(identifier))
            identifier = "/item:" + identifier;

        var process = Process.Start(new ProcessStartInfo(exePath, "\"" + assemblyPath + "\" /text /linenum " + identifier)
        {
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true
        });

        var projectFolder = Path.GetFullPath(Path.GetDirectoryName(assemblyPath) + "\\..\\..\\..").Replace("\\", "\\\\");
        projectFolder = Char.ToLower(projectFolder[0]) + projectFolder.Substring(1) + "\\\\";

        process.WaitForExit(10000);
        return string.Join(Environment.NewLine,
            Regex.Split(process.StandardOutput.ReadToEnd(), Environment.NewLine)
                .Where(l => !l.StartsWith("// Image base:"))
                .Select(l => l.Replace(projectFolder, ""))
        );
    }

    private static string GetPathToILDasm()
    {
        var path = Path.Combine(ToolLocationHelper.GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion.Version40), @"bin\NETFX 4.0 Tools\ildasm.exe");
        if (!File.Exists(path))
            path = path.Replace("v7.0", "v8.0");
        if (!File.Exists(path))
            Assert.Ignore("ILDasm could not be found");
        return path;
    }
}

#endif