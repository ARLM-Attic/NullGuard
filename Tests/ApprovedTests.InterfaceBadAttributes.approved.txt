
//  Microsoft (R) .NET Framework IL Disassembler.  Version 4.0.30319.1
//  Copyright (c) Microsoft Corporation.  All rights reserved.


// warning : THIS IS A PARTIAL DISASSEMBLY, NOT SUITABLE FOR RE-ASSEMBLING



.class interface private abstract auto ansi InterfaceBadAttributes
{
  .method public hidebysig newslot abstract virtual 
          instance void  MethodWithNoNullCheckOnParam(string arg) cil managed
  {
  } // end of method InterfaceBadAttributes::MethodWithNoNullCheckOnParam

  .method public hidebysig newslot specialname abstract virtual 
          instance string  get_PropertyWithNoNullCheckOnSet() cil managed
  {
  } // end of method InterfaceBadAttributes::get_PropertyWithNoNullCheckOnSet

  .method public hidebysig newslot specialname abstract virtual 
          instance void  set_PropertyWithNoNullCheckOnSet(string 'value') cil managed
  {
  } // end of method InterfaceBadAttributes::set_PropertyWithNoNullCheckOnSet

  .method public hidebysig newslot specialname abstract virtual 
          instance string  get_PropertyAllowsNullGetButDoesNotAllowNullSet() cil managed
  {
  } // end of method InterfaceBadAttributes::get_PropertyAllowsNullGetButDoesNotAllowNullSet

  .method public hidebysig newslot specialname abstract virtual 
          instance void  set_PropertyAllowsNullGetButDoesNotAllowNullSet(string 'value') cil managed
  {
  } // end of method InterfaceBadAttributes::set_PropertyAllowsNullGetButDoesNotAllowNullSet

  .method public hidebysig newslot abstract virtual 
          instance string  MethodAllowsNullReturnValue() cil managed
  {
  } // end of method InterfaceBadAttributes::MethodAllowsNullReturnValue

  .property instance string PropertyWithNoNullCheckOnSet()
  {
    .set instance void InterfaceBadAttributes::set_PropertyWithNoNullCheckOnSet(string)
    .get instance string InterfaceBadAttributes::get_PropertyWithNoNullCheckOnSet()
  } // end of property InterfaceBadAttributes::PropertyWithNoNullCheckOnSet
  .property instance string PropertyAllowsNullGetButDoesNotAllowNullSet()
  {
    .get instance string InterfaceBadAttributes::get_PropertyAllowsNullGetButDoesNotAllowNullSet()
    .set instance void InterfaceBadAttributes::set_PropertyAllowsNullGetButDoesNotAllowNullSet(string)
  } // end of property InterfaceBadAttributes::PropertyAllowsNullGetButDoesNotAllowNullSet
} // end of class InterfaceBadAttributes

// *********** DISASSEMBLY COMPLETE ***********************
